ipa-core/src/protocol/ipa_prf/boolean_ops/share_conversion_aby.rs

```diff
@@ -126,84 +128,113 @@ where
+            integer_add::<_, TwoHundredFiftySixBitOpStep, NC>(
+                ctx_mask_x.clone(),
+                record_id_x,
+                &sh_rs,
+                x,
+            )
+            .await?
+            .0,
+        );
+        vec_sh_s.push(sh_s);
+        vec_sh_r.push(sh_r);
+    }
 
     // validate before reveal
-    validator.validate().await?;
-
-    // this leaks information, but with negligible probability
-    let mut y = (m_ctx.role() != Role::H3).then(|| Vec::with_capacity(NC));
-    for i in 0..BITS {
-        let y_bit = partial_reveal(
-            m_ctx.narrow(&Step::RevealY(i)),
-            record_id,
-            Role::H3,
-            sh_y.get(i).unwrap(),
-        )
-        .await?;
-        match (&mut y, y_bit) {
-            (Some(y), Some(y_bit)) => y.push(y_bit),
-            (None, None) => (),
-            _ => unreachable!("inconsistent partial_reveal behavior"),
+    dzkp_validator.validate_chunk(ctx_counter).await?;
+
+    // output vec
+    // non vectorized Fp25519 shares
+    let mut results = Vec::<AdditiveShare<Fp25519, NP>>::with_capacity(NC * vec_sh_y.len());
+
+    for (i, ((sh_y, sh_r), sh_s)) in vec_sh_y.into_iter().zip(vec_sh_r).zip(vec_sh_s).enumerate() {
+        let record_id_x = first_record_id + i;
+        // this leaks information, but with negligible probability
+        let mut y = (m_ctx.role() != Role::H3).then(|| Vec::with_capacity(NC));
+        for i in 0..BITS {
+            let y_bit = partial_reveal(
+                m_ctx.narrow(&Step::RevealY(i)),
+                record_id_x,
+                Role::H3,
+                sh_y.get(i).unwrap(),
+            )
+            .await?;
+            match (&mut y, y_bit) {
+                (Some(y), Some(y_bit)) => y.push(y_bit),
+                (None, None) => (),
+                _ => unreachable!("inconsistent partial_reveal behavior"),
+            }
         }
-    }
 
-    let y = y.map(|y| {
-        Vec::<BA256>::transposed_from(y.as_slice().try_into().unwrap()).unwrap_infallible()
-    });
+        let y = y.map(|y| {
+            Vec::<BA256>::transposed_from(y.as_slice().try_into().unwrap()).unwrap_infallible()
+        });
 
-    let sh_r = Vec::<AdditiveShare<BA256>>::transposed_from(&sh_r)
-        .ok()
-        .expect("sh_r was constructed with the correct number of bits");
-    let sh_s = Vec::<AdditiveShare<BA256>>::transposed_from(&sh_s)
-        .ok()
-        .expect("sh_s was constructed with the correct number of bits");
+        let sh_r = Vec::<AdditiveShare<BA256>>::transposed_from(&sh_r)
+            .ok()
+            .expect("sh_r was constructed with the correct number of bits");
+        let sh_s = Vec::<AdditiveShare<BA256>>::transposed_from(&sh_s)
+            .ok()
+            .expect("sh_s was constructed with the correct number of bits");
 
-    output_shares::<_, NC, NP>(&m_ctx, &sh_r, &sh_s, y)
+        results.extend(output_shares::<_, NC, NP>(&m_ctx, &sh_r, &sh_s, y)?);
```

Comment:
`output_shares` is already pushing things onto a vector internally, so it would be a pretty simple change (and a likely inconsequential performance improvement) to pass `&mut results` into `output_shares` and save a little work.